### DISCUSSION TITLE
electron环境下撤销、重做等按钮功能无效如何解决

### DISCUSSION THREAD

**Discussion by rose8928** [opensumi/core#1791] [Q&A]
在electron环境下，撤销、重做按钮无效，查看源码是electron.undo事件，我又重写了这个事件，但是还是无效。
相当于是我重写了这个id为electron.undo事件，并且重写了一个他的菜单注册上去，但是单机对应菜单后，不执行相应的命令，我如果把electron.undo换成别的就是可以的，是命令不能被覆盖重写吗。那撤销按钮无效我该如何解决
![屏幕截图 2022-10-17 101627](https://user-images.githubusercontent.com/35814595/196075239-ef9e0fee-7f2f-4c25-94b2-74646168c2e4.png)
![屏幕截图 2022-10-17 101553](https://user-images.githubusercontent.com/35814595/196075242-a3d115ff-c068-46ee-bc2d-d0010d79976a.png)
![屏幕截图 2022-10-17 101416](https://user-images.githubusercontent.com/35814595/196075246-93ff7347-c8ca-4804-ab2e-a941bbd3139f.png)

**Comment by yantze** [comment]
1. execute: () => {} 这里是最后的执行地方，你这里没有做任何 undo 的行为。

2. 另外看看一下源码，需要加字段 nativeRole：https://github.com/opensumi/core/blob/50b22026968bd2119b877621ba5aad955760c7cc/packages/electron-basic/src/browser/index.ts#L150

**Comment by rose8928** [comment]
我只是测试一下execute能否执行，所以没写undo行为，随便写了一个行为，但也不执行。
请问一下nativeRole的作用是什么，如何使用的。好像是非必传字段。

**Reply by erha19** [reply]
这里的 nativeRole 会在 Electron 生成菜单时转化为对应的 `Menu#role` 属性，参考 https://www.electronjs.org/docs/latest/api/menu

**Comment by rose8928** [comment]
@yantze @erha19 我在git把ide-electron克隆，没有改任何代码，运行了一下，undo和redo也是无效的，是没有实现吗，能教我一下怎么实现undo和redo吗

**Reply by yantze** [reply]
我们本地测试多个系统都是正常的执行 unde / redo，你可以通过下面的步骤来逐步排查：
1. 看看是不是上一次的编译产物没有删除
2. 你本地项目断点一下这里有没有被 undo 触发 https://github.com/opensumi/core/blob/6663c197e14f1eccecc2ea23e88d68ff62c19b49/packages/electron-basic/src/browser/index.ts#L198

**Comment by rose8928** [comment]
@yantze ，触发了，我在上面加了一行this.messageService.info()，能够显示信息，但是不执行undo操作，
请问一下runNativeRoleAction的实现在哪个文件
我再看看，为啥我这里ide-electron的undo就不行啊，我在试一下。web的就可以，electron不行，哎

**Comment by bytemain** [comment]
你说的 undo 操作是在编辑器中 undo 还是你自定义的某个页面进行 undo？
在编辑器中的 undo 是由 monaco 自己实现的。

**Reply by rose8928** [reply]
> 你说的 undo 操作是在编辑器中 undo 还是你自定义的某个页面进行 undo？ 在编辑器中的 undo 是由 monaco 自己实现的。

就是js文件或者json文件的undo或者redo，快捷键ctrl+z是可以用的，但是点击上面的编辑菜单里的撤销和重做不生效。

**Comment by yantze** [comment]
怎么感觉像是其它软件设置全局快捷键导致的遗留问题，所以重启电脑一下试试
